node/core/provisioner/src/lib.rs

```diff
@@ -0,0 +1,362 @@
+						self.signed_bitfields.clone(),
+						self.backed_candidates.clone(),
+						sender,
+						self.sender.clone(),
+					)
+					.await?
+				}
+				ToJob::Provisioner(RequestBlockAuthorshipData(_, sender)) => {
+					self.provisionable_data_channels.push(sender)
+				}
+				ToJob::Provisioner(ProvisionableData(data)) => {
+					for channel in self.provisionable_data_channels.iter_mut() {
+						// REVIEW: the try operator here breaks the run loop if any receiver ever unexpectedly
+						// closes their channel. Is that desired?
+						channel.send(data.clone()).await?;
+					}
+					self.note_provisionable_data(data);
+				}
+				ToJob::Stop => break,
+			}
+		}
+
+		Ok(())
+	}
+
+	fn note_provisionable_data(&mut self, provisionable_data: ProvisionableData) {
+		match provisionable_data {
+			ProvisionableData::Bitfield(_, signed_bitfield) => {
+				self.signed_bitfields.push(signed_bitfield)
+			}
+			ProvisionableData::BackedCandidate(backed_candidate) => {
+				self.backed_candidates.push(backed_candidate)
+			}
+			_ => {}
+		}
+	}
+}
+
+// The provisioner is the subsystem best suited to choosing which specific
+// backed candidates and availability bitfields should be assembled into the
+// block. To engage this functionality, a
+// `ProvisionerMessage::RequestInherentData` is sent; the response is a set of
+// non-conflicting candidates and the appropriate bitfields. Non-conflicting
+// means that there are never two distinct parachain candidates included for
+// the same parachain and that new parachain candidates cannot be included
+// until the previous one either gets declared available or expired.
+//
+// The main complication here is going to be around handling
+// occupied-core-assumptions. We might have candidates that are only
+// includable when some bitfields are included. And we might have candidates
+// that are not includable when certain bitfields are included.
+//
+// When we're choosing bitfields to include, the rule should be simple:
+// maximize availability. So basically, include all bitfields. And then
+// choose a coherent set of candidates along with that.
+async fn send_inherent_data(
+	relay_parent: Hash,
+	signed_bitfields: Vec<SignedAvailabilityBitfield>,
+	mut backed_candidates: Vec<BackedCandidate>,
+	return_sender: oneshot::Sender<ProvisionerInherentData>,
+	mut from_job: mpsc::Sender<FromJob>,
+) -> Result<(), Error> {
+	let availability_cores = match request_availability_cores(relay_parent, &mut from_job)
+		.await?
+		.await?
+	{
+		Ok(cores) => cores,
+		Err(runtime_err) => {
+			// Don't take down the node on runtime API errors.
+			log::warn!(target: "provisioner", "Encountered a runtime API error: {:?}", runtime_err);
+			return Ok(());
+		}
+	};
+
+	// availability cores indexed by their para_id for ease of use
+	let cores_by_id: HashMap<_, _> = availability_cores
+		.iter()
+		.enumerate()
+		.filter_map(|(idx, core_state)| Some((core_state.para_id()?, (idx, core_state))))
```

Comment:
This is only giving the "top-level" `ParaId` and not the `ParaId` that is supplied in the `next_up_on...`.

I recommend doing a bit of code restructuring:
  1. Select bitfields
  2. Do a preprocessing loop on the `Vec<CoreState>`. Replace all `OccupiedCore`s with the `ScheduledCore` from `next_up_on_available` or `next_up_on_time_out` as appropriate. If neither options are appropriate, leave as occupied.
  3. Loop over the modified `Vec<CoreState>` and include one candidate per `ScheduledCore`.

Reply:
Also, the preprocessing loop (step 2) should tag each `CoreState` with an `OccupiedCoreAssumption` depending on what happened to the core. If the core was already `Scheduled` or `Free`, use `OccupiedCoreAssumption::Freed`. If the core was made available then use `OccupiedCoreAssumption::Included`. If it was timed out, then use `OccupiedCoreAssumption::TimedOut`. If the core remains occupied, it doesn't matter what we put here since we won't include any candidate for that core. 

So instead of the preproccessing outputting a  `Vec<CoreState>`, we'll leave out occupied / unscheduled cores and create a `Vec<(CoreIndex, ScheduledCore, OccupiedCoreAssumption)>`. The `CoreIndex` might be unnecessary. If so, it should be left out.

Then, the candidate selection step (3) will select a candidate for each scheduled core whose `validation_data_hash` matches the `local_validation_data` under that `OccupiedCoreAssumption`. If there is none, then we don't select it.

Reply:
fb42634

Reply:
That commit does not implement what is described here.

Reply:
Ah--you're right. I'll add that constraint with the `validation_data_hash`.